### ISSUE TITLE
[question] Is it possible to run a pre-build step only once for all variants 

### ISSUE THREAD

**Issue by maitrey** [conan-io/conan#11807]
Dear Conan Folks, 
I am back again with another question. 
We use cmake-3.21.1(MinGw as makefile generator), Windows and conan-1.50.
Our builds needs:
- Generation of code from models
This is a pre-build step for which we are using add_custom_target and providing dependencies to the main target such that the pre-build step is done first.
I have about 6 variants. I would like to run the pre-build step for generating code  only once so that it can be reused for all the variants for generating different executables.
Is it possible to do so?

- [ x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by lasote** [comment]
Hi, maybe you can create a `build require` for the generation from models.

**Comment by maitrey** [comment]
Hi Lasote, Thanks for your reply. Do you mean I should split the generation of source code into a seperate package and indicate it as build requires? But the main reason for having them together is that  : 

- We have the models, we generate code, test them on target compiler and when all are passing we ship the generated code along with the compiled code. 
Moreover, too many packages also adds up to the vertical layer of dependencies leading to re-releases of dependencies upwards.

**Comment by memsharded** [comment]
Hi @maitrey 

It is not clear what is your intention and constraints. Why do you want to generate all variants? Code generation is typically part of the ``build()`` step, specially if the generated code is different for each variant. Why do you need to generate them all in a single step, if only one at a time can be used? How are you controlling those variants, are those Conan options? What is the meaning of packaging/shipping different source code variants with just 1 variant of the binaries?

**Comment by maitrey** [comment]
Sorry perhaps I was not making myself clear. So the code generation ideally shall happen once but compiled  into 6 different variants(different compilers,flags  defined in profiles) . Since the code generation happens in the build step of the package , when I run conan create it generates also 6 times which is kind of unnecessary. I wanted to ask if there is any possible way to eliminate running the generation 6 times? What would help me is:
- **Generate the code ONLY once from models as a pre-build step** 
- Use the generated code and compile it in 6 different variants
- Ship the generated code + executables

**Comment by memsharded** [comment]
I am sorry, but still not fully clear. Are the 6 different variants, 6 different binary packages with 6 different ``package_ids``? Are those variants based on ``settings`` and ``options``? Do you want to have each one of the 6 different binary packages include a copy of the generated source code for all the different 6 variants, even if the package binary contains only 1 variant?

Shipping to the final usage/user/deployment/customer is a different thing, but Conan packages will contain different binaries for each different variants. Those binaries still belong to the same "package", but they scale horizontally, Conan manages them automatically, so it doesn't add complexity to the management. After development time, you might want to collect everything that you are shipping to your production servers or to your customers from Conan packages. What do you mean with ``Ship``? How are you shipping your final product?

**Comment by maitrey** [comment]
There are 6 different variants with 6 different package_ids . 6 different variants are based on different profiles .
- repository contains the model. When I say conan create in the build method it generates the code and builds/tests it. This happens 6 times which is unecessary. Generated code is variant independent. The way it is compiled differs for the 6 different variants.
Input model --->Output Generated Code ----> Generated Code is compiled in 6 different variants
So my question was can I minimize the generated code to be generated only once and not 6 times as this is NOT needed.
By ship means deploying it it in artifactory and controlling it with access management.

**Comment by memsharded** [comment]
> repository contains the model. When I say conan create in the build method it generates the code and builds/tests it. This happens 6 times which is unecessary. Generated code is variant independent. The way it is compiled differs for the 6 different variants.

Now I start to understand, thanks! :)

If the code is the same for the different variants, why not using the ``source()`` method to generate it? This is basically the purpose of this method, to download/clone/prepare the source code that will be compiled in ``build()``, as long as it is common code for all the ``build()`` operations, because it runs only once, and then it is reused for all the subsequent builds. 

That would work for builds in the same machine, of course. If this step is to be done in a distributed build, then the only way is to use a Conan package that contains such generated code, and is reused by the other package.

**Comment by maitrey** [comment]
The reason for using build() method was I could use some modules from cmake like find_matlab in order to find the location of matlab exe. Offcourse I can do the same using python but because of working in an enterprise, access is not given to query windows registry keys . However, that is not a conan issue. I am glad to know that we can use the source method for such use-cases. Thanks for your reply.

**Comment by memsharded** [comment]
Thanks for following up. If the ``source()`` method is the response, then I am closing the issue as solved, but please do not hesitate to re-open or file a new ticket if you have further questions or need help with something else. Cheers!
